src/utils/finance/source.py

```diff
@@ -0,0 +1,266 @@
+            for duplicate_column in duplicate_column_list:
+                standard_data_table[duplicate_column] = standard_data_table[base_column]
+        return standard_data_table
+
+    def _add_new_columns(self, standard_data_table: pd.DataFrame) -> pd.DataFrame:
+        """Add new blank columns TODO: is this ever really needed?"""
+        for column in self.new_empty_columns:
+            standard_data_table[column] = None
+        return standard_data_table
+
+    def _add_state_code(self, standard_data_table: pd.DataFrame) -> pd.DataFrame:
+        """Add state code to relevent columns"""
+        for column in self.state_code_columns:
+            standard_data_table[column] = self.state_code
+        return standard_data_table
+
+    def standardize_schema(self, raw_data_table: pd.DataFrame) -> pd.DataFrame:
+        """Rename columns, remove unused columns, and add missing columns
+
+        Args:
+            raw_data_table: single dataframe with raw column names
+        """
+        standard_column_name_table = self._rename_columns(raw_data_table)
+        standard_relevant_column_table = self._drop_unused_columns(
+            standard_column_name_table
+        )
+        standard_relevant_column_table = self._add_state_code(
+            standard_relevant_column_table
+        )
+        standard_relevant_column_table = self._add_new_columns(
+            standard_relevant_column_table
+        )
+        standard_schema_table = self._add_duplicate_columns(
+            standard_relevant_column_table
+        )
+
+        return standard_schema_table
+
+
+class DataStandardizer:
+    """Ensures data conforms to expected types"""
+
+    def __init__(self, config_handler: ConfigHandler | None = None) -> None:
+        """Create a new DataStandardizer object
+
+        Args:
+            config_handler: ConfigHandler, documented # TODO
+        """
+        self.enum_mapper = config_handler.enum_mapper
+        self.column_to_date_format = config_handler.column_to_date_format
+
+    def _standardize_enums(self, standard_schema_table: pd.DataFrame) -> pd.DataFrame:
+        """Rename entity type columns"""
+        for column_name, column_enum_map in self.enum_mapper.items():
+            if column_name not in standard_schema_table.columns:
+                raise ValueError(f"Provided enum: {column_name} not in table")
+            # map each value in the table's column according to the
+            # provided enum mapper
+            standard_schema_table[column_name] = standard_schema_table[column_name].map(
+                column_enum_map
+            )
+        return standard_schema_table
+
+    def _standardize_column_to_date_format(
+        self, standard_schema_table: pd.DataFrame
+    ) -> pd.DataFrame:
+        for date_column in self.column_to_date_format:
+            na_mask = standard_schema_table[date_column].isna()
+            temp_column = f"tmp-{date_column}"
```

Comment:
I think just putting the formatted dates in a series would be simpler here. You can just build out a series called like `formatted_dates` or whatever, and then do `standard_scema_table[date_column] = formatted_dates`, right?

Reply:
Or wait...I don't actually understand why you need to mask the NaNs at all. I think pd.to_datetime handles NaNs fine when using errors="coerce"? Unless I guess you're trying to distinguish between NaN and NaT?

Reply:
This brings back memories of a very difficult bug. There was inconsistent behavoir when NaNs were mixed that I believe was due to this: https://github.com/pandas-dev/pandas/issues/58924. It looks like it's resolved now, but not until pandas 3.0. I think it makes sense to keep this as is since it works.

Reply:
Fascinating. Sounds good.